docs/experiments/extract-mode.md

```diff
@@ -0,0 +1,103 @@
+
+## Example of extract mode
+
+Without extract mode, writing rules to validate template, Markdown or configuration files which contain code in another language can be burdensome and require significant rule duplication.
+
+Let's take the following Bash rule as an example (a simplified version of [our `curl-eval` rule](https://github.com/returntocorp/semgrep-rules/blob/release/bash/curl/security/curl-eval.yaml)):
+
+```yaml
+rules:
+  - id: curl-eval
+    severity: WARNING
+    languages:
+      - bash
+    message: Evaluating data from a `curl` command is unsafe.
+    mode: taint
+    pattern-sources:
+      - pattern: |
+          $(curl ...)
+      - pattern: |
+          `curl ...`
+    pattern-sinks:
+      - pattern: eval ...
+```
+
+Usually, this rule would only be run against Bash files. However, our project might contain Dockerfiles or Python scripts which invoke Bash commands&mdash;without an extract mode rule Semgrep won't know about any contained Bash, and therefore, will **not** run any Bash rules against those contained commands.
+
+However, with extract mode we can provide Semgrep with instructions on how to extract any Bash commands we might be using in a Docker `RUN` instruction or as an argument to Python's `os.system` standard library function.
+
+```yaml
+rules:
+  - id: extract-docker-run-to-bash
+    mode: extract
+    languages:
+      - dockerfile
+    pattern: RUN $...CMD
+    extract: $...CMD
+    dest-language: bash
+  - id: extract-python-os-system-to-bash
+    mode: extract
+    languages:
+      - python
+    pattern: os.system("$CMD")
+    extract: $CMD
+    dest-language: bash
+```
+
+By adding the extract mode rules show above, Semgrep can match the Bash contained in the below Python file, and will report the contained Bash as  matching against the `curl-eval` rule.
+
+```python
+from os import system
+
+if system('eval `curl -s "http://www.very-secure-website.net"`'):
+    print("Command failed!")
+else:
+    print("Success")
+```
+
+Likewise, if our query included a Dockerfile with an equivalent Bash command, such as the example below, Semgrep would report the contained Bash as matching against the `curl-eval` rule.
+
+```dockerfile
+FROM fedora
+RUN dnf install -y unzip zip curl which
+RUN eval `curl -s "http://www.very-secure-website.net"`
+```
+
+## Extract mode rule schema
+
+Extract mode rules **require** the following [usual Semgrep rule keys](../../writing-rules/rule-syntax/#required):
+  - `id`
+  - `languages`
+  - One of `pattern`, `patterns`, `pattern-either`, or `pattern-regex`
+
+Extract mode rules **also require** two additional fields:
+  - `extract`
+  - `dest-language`
+
+The extract mode specific fields are further explained in two sections below.
+
+### `extract`
+
+The `extract` key is required when in extract mode. The value must be a metavariable appearing in your pattern(s), and the value bound to that metavariable will be extracted.
```

Comment:
Blocking: 
- Use active voice - is extracted - Semgrep extracts - what do we mean by extracting it actually? Detected maybe?
- Avoid [will](https://developers.google.com/style/word-list?hl=en#will) when possible. Documentation needs to be written mostly in the present simple tense: See will in the Google developer style guide.

```suggestion
The `extract` key is required when in extract mode. The value must be a metavariable appearing in your pattern(s), and Semgrep extracts the value bound to that metavariable.
```

Reply:
I wasn't sure how much to talk about the actual mechanics of what we do for a page like this, but the "extraction" essentially comprises the process of reading the text bound to the metavariable from the original file, writing it to a temporary file, and then running rules for the destination language against that file.

It is unclear to me if that description is more or less helpful than the admittedly nebulous "extracts".

The suggestion you have seems good to me wrt your other points.

Reply:
[migrated to other branch]